inference-engine/src/inference_engine/include/openvino/runtime/async_infer_queue.hpp

```diff
@@ -0,0 +1,120 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/**
+ * @brief A header file that provides wrapper classes for infer requests and callbacks.
+ *
+ * @file openvino/runtime/async_infer_queue.hpp
+ */
+#pragma once
+
+#include <condition_variable>
+#include <map>
+#include <mutex>
+#include <queue>
+#include <string>
+#include <vector>
+
+#include "openvino/runtime/common.hpp"
+#include "openvino/runtime/executable_network.hpp"
+#include "openvino/runtime/infer_request.hpp"
+#include "openvino/runtime/tensor.hpp"
+
+namespace ov {
+namespace runtime {
+
+/**
+ * @brief This is an interface of asynchronous infer request
+ *
+ * It can throw exceptions safely for the application, where it is properly handled.
+ * 
+ * @note AsyncInferQueue provides simple and unifed interface for multiple asynchronous
+ * calls to OpenVINO's InferRequests.
+ */
+class OPENVINO_RUNTIME_API AsyncInferQueue {
+public:
+    /**
+     * @brief Constructor
+     */
+    AsyncInferQueue(ov::runtime::ExecutableNetwork net, size_t jobs = 0);
```

Comment:
Can it accept `std::vector<ov::InferRequest>` so we can generalize it for several plugins?

Reply:
Of course, I can provide second constructor which will extract number of jobs from vector's size and create unique ids based on that. Can you provide some more info about such use cases?

However the promise is that all `InferRequest` must be created from the same `ExecutableNetwork`. Without that many different networks can be run and setting one universal callback could be difficult for the user.

Reply:
After deeper analysis this is a very big feature request that needs to be consulted with architecture forum first.